### ISSUE TITLE
Asked to login again for few links opened in new tab

### ISSUE THREAD

**Issue by shvmbisht** [backstage/backstage#8462]
TL;DR When a logged-in user opens the app in the new link, the user gets the login screen again in the newly opened tab, and  in the Developer tools network TAB, they get `TimeoutError: Timeout awaiting 'request' for 3500ms` as a response for the API `/api/auth/oidc/refresh?optional&env=development`, same error occurred for both GitLab and oidc. This error is not very consistent.

## Expected Behavior
Users should not be asked to log in again if they are already login.

## Current Behavior
When a logged-in user tries to open a link in a new tab, the user is asked to log in again, this happens for a few links and it works fine for other links in the same platform.

## Steps to Reproduce

1. PFA video and screenshots

https://user-images.githubusercontent.com/19923220/146131531-14da3e1f-d7ea-44c1-9302-c3b69806dd37.mp4

![scrnli_15_12_2021_10-56-07](https://user-images.githubusercontent.com/19923220/146128656-ba425cb5-ab99-4c13-8514-01673f10ddf2.png)
![scrnli_15_12_2021_10-56-41](https://user-images.githubusercontent.com/19923220/146128680-4504329f-090f-4e5e-b9c7-cdb297c1481c.png)

## Your Environment

- NodeJS Version (v14): v14.17.5
- Operating System and Version (e.g. Ubuntu 14.04): MacOs Big Sur
- Browser Information: chrome

**Comment by Rugvip** [comment]
I believe this might be because of a race that happens when the backing authentication provider invalidates refresh tokens on use. So if you for example open two tabs at once, both are gonna trigger an auth refresh, and if the second one is opened before the first one has had time to do a refresh, it will use the same refresh token and then fail to sign-in as the refresh token has been invalidated.

One way forward that I see if ^ is indeed the issue is to re-use the existing session for any new tabs that are opened. Something like a bit of `postMessage`ing between tabs to re-use sessions rather than always refreshing them from scratch. Happy to bounce other ideas though as that could be a bit tricky to execute safely. Other options could be to try to do something smart in the auth-backend, or to switch to a more conventional session store, at least as an available option.

**Comment by mufaddal7** [comment]
@Rugvip Do we have any future plans/proposals to move to a more conventional session store?

**Comment by Rugvip** [comment]
@mufaddal7 No nothing planned atm, I could help guide the work if it's something you want to pick up though 😁

**Comment by mufaddal7** [comment]
@Rugvip Sure, we can pick this. If we have a good roadmap, it will be easy to implement. Any thoughts on using redux for state management? Refresh mechanism can be managed more efficiently via redux

**Comment by freben** [comment]
Redux hasn't been a pattern elsewhere and we're not particularly eager to start using it :) Essentially, an adopter is free to use any state management they see fit of course, but for core features, we have held back on adding any specific one. If you could make do with useState and contexts etc, it would probably be best.

**Comment by mufaddal7** [comment]
@freben Context API is good and lightweight, we can do something in line with that. Would look forward to working on this. As @Rugvip suggested if you guys can give an outline/plan, that would be a good place to start.

**Comment by Rugvip** [comment]
I'm not sure we want to completely move over to a shared session store for all backends at this point, since that would be a pretty big shift. By that I mean for example catalog backend always be able to access secrets written by the auth backend, as those my not be deployed together on the same node. Adding this potentially breaks down some of our isolation between backends as well, so would need to consider that more carefully.

What I think could make sense for now is to refactor the refresh token storage, so that instead of always storing it in a cookie, you can instead also store it in the existing [session store](https://github.com/backstage/backstage/blob/a85d9fe831603bb82fcb6807236d8e9dd398d18e/plugins/auth-backend/src/service/router.ts#L68) of the auth-backend. I think the refactor would be something along the lines of "if a session store is configured, store refresh tokens there". It may be better to take that on post #6245 though, as having the auth provider implementations be more uniform makes it easier to do broader changes.

Just switching to the existing session storage wouldn't really solve this problem though, as it's still stored in a cookie, and that's where I'm thinking it makes sense to pick up this work. That is allowing a persistent session store to be configured rather than the existing in-memory one. There are [plenty](https://www.npmjs.com/package/express-session#compatible-session-stores) of possible options here but in the end I think we at least would want something that works with our existing DB infra and that is configured through our app-config, whatever is the cleanest way to get there :grin: Do you think it makes sense to start there?

**Comment by NishkarshRaj** [comment]
@Rugvip @freben we are still facing issues similar to what @shvmbisht had raised.

Given the fact that #6245 is completed, can we please re-open this?

We are happy to help contribute to this effort.
